mantidproject/mantid
# New algorithm: ReflectometryCorrectDetectorAngle

**Description of work.**

This PR introduces a new algorithm `ReflectometryCorrectDetectorAngle` which can be used to adjust the specular angle of a line detector.

**Report to:** [nobody].

**To test:**

1. The script below uses some files from the unit test data directories. It exemplifies the three different modes of operation of the algorithm. Check that the script runs, and then try to give the algorithm bogus properties. Can you make it fail?

```python
import numpy as np
reflected = Load('ILL/D17/317370.nxs')
reflected = ExtractMonitors(reflected, DetectorWorkspace='reflected')

# Set detector centre angle to some value
corrected1 = ReflectometryCorrectDetectorAngle(reflected,
    'detector', TwoTheta=2.3)
# The centre is in the middle of pixels 127 and 128
assert(corrected1.spectrumInfo().twoTheta(127) > np.deg2rad(2.3))
assert(corrected1.spectrumInfo().twoTheta(128) < np.deg2rad(2.3))

# Set the angle of certain pixel (workspace index)
# This should be equivalent to above
corrected2 = ReflectometryCorrectDetectorAngle(reflected,
    'detector', LinePosition=127.5, PixelSize=0.001195, TwoTheta=2.3)
assert(corrected1.spectrumInfo().twoTheta(0) == corrected2.spectrumInfo().twoTheta(0))
# An arbitrary pixel
corrected3 = ReflectometryCorrectDetectorAngle(reflected,
    'detector', LinePosition=22, PixelSize=0.001195, TwoTheta=4.2)
print(np.rad2deg(corrected3.spectrumInfo().twoTheta(22)))
print(4.2)

# Correct angle using a direct beam reference
LoadILLReflectometry('ILL/D17/317369.nxs', OutputWorkspace='direct', OutputBeamPosition='direct_beam_pos')
direct = ExtractMonitors('direct', DetectorWorkspace='direct')
direct_beam_pos = mtd['direct_beam_pos'].cell('PeakCentre', 0)
print(direct_beam_pos)
# The direct beam is not quite at 0 degrees.
print(np.rad2deg(direct.spectrumInfo().twoTheta(int(direct_beam_pos))))
# Reflected detector angle needs to be calibrated accordingly
corrected4 = ReflectometryCorrectDetectorAngle(reflected,
    'detector', PixelSize=0.001195, DirectLinePosition=direct_beam_pos,
    DirectBeamWorkspace='direct')
print(np.rad2deg(reflected.spectrumInfo().twoTheta(0)))
print(np.rad2deg(corrected4.spectrumInfo().twoTheta(0)))
```

2. Compare the algorithm with `LoadILLReflectometry`.

```python
import numpy as np
# Without direct beam calibration
reflected = Load('ILL/D17/317370.nxs')
reflected = ExtractMonitors(reflected, DetectorWorkspace='reflected')
corrected1 = ReflectometryCorrectDetectorAngle(reflected,
    'detector', TwoTheta=2.3)
corrected2 = LoadILLReflectometry('ILL/D17/317370.nxs', BeamCentre=127.5, BraggAngle=2.3 / 2.)
assert(corrected1.spectrumInfo().twoTheta(0) == corrected2.spectrumInfo().twoTheta(0))

# With direct beam calibration
LoadILLReflectometry('ILL/D17/317369.nxs', OutputWorkspace='direct', OutputBeamPosition='direct_beam_pos')
direct = ExtractMonitors('direct', DetectorWorkspace='direct')
direct_beam_pos = mtd['direct_beam_pos'].cell('PeakCentre', 0)
corrected3 = ReflectometryCorrectDetectorAngle(reflected,
    'detector', PixelSize=0.001195, DirectLinePosition=direct_beam_pos,
    DirectBeamWorkspace='direct')
corrected4 = LoadILLReflectometry('ILL/D17/317370.nxs', BeamCentre=127.5, DirectBeamPosition='direct_beam_pos')
assert(corrected3.spectrumInfo().twoTheta(0) == corrected4.spectrumInfo().twoTheta(0))
```

Fixes partially #22734.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
